File: 7.二叉树/8.对称二叉树/100.相同的树.cpp
/*
 * @lc app=leetcode.cn id=100 lang=cpp
 *
 * [100] 相同的树
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        return dfs(p, q);
    }
private:
    bool dfs(TreeNode* n1, TreeNode* n2)
    {
        if (!n1 || !n2)
        {
            return n1 == n2;
        }
        return n1->val == n2->val && dfs(n1->left, n2->left) && dfs(n1->right, n2->right);
    }
};
// @lc code=end
class Solution {
    public:
        bool isSameTree(TreeNode* p, TreeNode* q) {
            return dfs(p, q);
        }
    private:
        bool dfs(TreeNode* n1, TreeNode* n2)
        {
            if (!n1 && n2) return false;
            else if (n1 && !n2) return false;
            else if (!n1 && !n2) return true;
            else if (n1->val != n2->val) return false;
            bool l =  dfs(n1->left, n2->left);
            bool r = dfs(n1->right, n2->right);
            return l & r;
        }
    };

File: 面经150/面经150p1/9.二叉树/100.相同的树.py
#
# @lc app=leetcode.cn id=100 lang=python3
#
# [100] 相同的树
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def dfs(p, q):
            if not p and not q: return True
            elif (p and not q) or (q and not p): return False
            return p.val == q.val and dfs(p.left, q.left) and dfs(p.right, q.right)
        return dfs(p, q)
# @lc code=end

